### DISCUSSION TITLE
How does Cohesive zone actually works? for modelling stress with faults

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#18240] [Q&A Modules: Solid mechanics] [answered]
Hi guys,

The documentation and examples for CZM are quite insufficient to me, I'm not sure how to implement it to my problem.

https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/CohesiveZoneMaster/

Let's say the example below.

the objective of the simulation is to look into stress/strain change over mining steps (similar to coal mining problem)

![czm](https://user-images.githubusercontent.com/79562439/124224824-b7579d00-db49-11eb-84d3-310ed75b99f9.PNG)

Block = blockID saperation
Red = faults OR cohesive zone (cohesive element in Abaqus), as I read from 'CohesiveZoneMaster', `BreakMeshByBlockGenerator` will duplicate the note at the blocks above and below the fault line (red). Can I basically just import a similar Abaqus mesh without having to use `BreakMeshByBlockGenerator`?

Then I am not sure how does it work, will `CZMInterfaceKernel` automatically pick up these fault lines?

as I see from the example, I only see this ...

```
[Modules/TensorMechanics/CohesiveZoneMaster]
  [./czm1]
    boundary = 'interface'
    displacements = 'disp_x disp_y disp_z'
  [../]
[]
```

Then the properties of CohesiveZone is later defined under `[Materials]` block

```
  [./czm_3dc]
    type = SalehaniIrani3DCTraction
    boundary = 'interface'
    normal_gap_at_maximum_normal_traction = 1
    tangential_gap_at_maximum_shear_traction = 0.5
    maximum_normal_traction = 100
    maximum_shear_traction = 70
    displacements = 'disp_x disp_y disp_z'
  [../]
```

and that's pretty much it, I am not sure if MOOSE's czm is applicable to my problem, would be nice if we have cohesive elements like Abaqus instead, so we basically can identify the faults from an input mesh.

It would be nice if someone could shed some light. Thank you

Kind regards,
Traiwit

**Comment by Traiwit** [comment]
Just wondering if anyone has ever implemented CZM with MOOSE at all?

**Comment by hugary1995** [comment]
I'm not sure what you mean by "like Abaqus". Moose can definitely import a mesh from Abaqus .inp file though. Then BreakMeshByBlockGenerator should work.

@arovinelli knows more about czm in moose.

**Reply by Traiwit** [reply]
Thanks for your reply @hugary1995.

As far as I know, Abaqus use cohesive elements but MOOSE use cohesive interfaces. Hence, I'm not sure if I import mesh with cohesive elements from Abaqus, will MOOSE czm automatically recognise it as an interface OR I have to use `BreakMeshByBlockGenerator` in order to generate interfaces and remove cohesive elements from the mesh.

@arovinelli I'm also following your finite strain CZM, looking forward to using it!

**Reply by hugary1995** [reply]
I see. I think you'll have to remove those cohesive elements in the Abaqus mesh.

**Reply by arovinelli** [reply]
@Traiwit It is definitely applicable to your problem as it's very similar to what I use it for.
The "problem" is setting up the mesh correctly. Unfortunately you can't use an ABAQUS mesh with cohesive elements directly for this problem. You will need to remove the ABAQUS cohesive elements, assign different "block_ids" to the different regions of your mesh around which you want cohesive interfaces, you could use cubit to achieve this. Then in your input file use break mesh by block generator to split the mesh. It will create a boundary called `interface` around all the blocks. That is the boundary representing the cohesive zone. 
So ideally you should start with a mesh that looks like the figure on the left.
![Screen Shot 2021-07-17 at 9 26 04 AM](https://user-images.githubusercontent.com/34654803/126040147-194507b7-8494-4bf1-b296-3292b69398aa.png)
The cohesive interface, (teh boundary named `interface` in MOOSE) will be like the on on the right.
When you specify `Modules/TensorMechanics/CohesiveZoneMaster/cmz1/boundary=interface` will tell MOOSE to add cohesive interface kernels on the boundary called `interface`, that's how the system work, no much more to do. Finally, in the materials you also specify the `boundary=interface` .  Nothing more is needed to make the simulation run.
Hope this helps, if you have more questions just let me know

**Reply by Traiwit** [reply]
Hi @arovinelli Thank for your reply!

I've tried to work with this testing mesh (attached), there are 2 blocks but for some reason, `BreakMeshByBlockGenerator` doesn't seem to work, it does not create an interface between these 2 blocks.

```
[mesh]
  type = FileMeshGenerator
   file = converting_in.e
   construct_side_list_from_node_list=true
   []
        [./breakmesh]
          input = mesh
          type = BreakMeshByBlockGenerator
          split_interface=true
          show_info = true
        [../]
  show_info = true
  allow_renumbering=false
```

Not sure if I miss anything. is there a limitation that I'm not aware of? for example, predefined duplicated nodes etc...

Kind regards,
Traiwi


![2021-07-21_23-50](https://user-images.githubusercontent.com/79562439/126499822-922d7f31-94a7-49b5-94a0-1a7aa16c9582.png)


[mesh_CZM_test.zip](https://github.com/idaholab/moose/files/6856203/mesh_CZM_test.zip)

**Reply by Traiwit** [reply]
Also another question, how would you work around with such a fault (attached figure), where the CZM is right in the middle of the geometry

![image](https://user-images.githubusercontent.com/79562439/126500271-ba3a66fd-3aac-4657-a633-685107bdad2e.png)

**Reply by arovinelli** [reply]
> Hi @arovinelli Thank for your reply!
> 
> I've tried to work with this testing mesh (attached), there are 2 blocks but for some reason, `BreakMeshByBlockGenerator` doesn't seem to work, it does not create an interface between these 2 blocks.
> 
> ```
> [mesh]
>   type = FileMeshGenerator
>    file = converting_in.e
>    construct_side_list_from_node_list=true
>    []
>         [./breakmesh]
>           input = mesh
>           type = BreakMeshByBlockGenerator
>           split_interface=true
>           show_info = true
>         [../]
>   show_info = true
>   allow_renumbering=false
> ```
> 
> Not sure if I miss anything. is there a limitation that I'm not aware of? for example, predefined duplicated nodes etc...
> 
> Kind regards,
> Traiwi
> 
> ![2021-07-21_23-50](https://user-images.githubusercontent.com/79562439/126499822-922d7f31-94a7-49b5-94a0-1a7aa16c9582.png)
> 
> [mesh_CZM_test.zip](https://github.com/idaholab/moose/files/6856203/mesh_CZM_test.zip)

yes the mesh should not already have duplicated nodes, otherwise the algorithm will not find neighbors. That is exactly what is happening in your case. 

For your second question is a bit trickier but the same principle apply. One way of doing it is to use the BreakMeshByBlockGenerator `block_pair` option see [here](https://mooseframework.inl.gov/source/meshgenerators/BreakMeshByBlockGenerator.html).  That would require careful design of domains id.

I guess what we should really do in your case is to implement a join mesh algorithm such that you could start from a mesh from which you have removed cohesive elements... If you fill this might be an important tool could you open a ticket requesting for it?

**Reply by jiangwen84** [reply]
> > Hi @arovinelli Thank for your reply!
> > I've tried to work with this testing mesh (attached), there are 2 blocks but for some reason, `BreakMeshByBlockGenerator` doesn't seem to work, it does not create an interface between these 2 blocks.
> > ```
> > [mesh]
> >   type = FileMeshGenerator
> >    file = converting_in.e
> >    construct_side_list_from_node_list=true
> >    []
> >         [./breakmesh]
> >           input = mesh
> >           type = BreakMeshByBlockGenerator
> >           split_interface=true
> >           show_info = true
> >         [../]
> >   show_info = true
> >   allow_renumbering=false
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Not sure if I miss anything. is there a limitation that I'm not aware of? for example, predefined duplicated nodes etc...
> > Kind regards,
> > Traiwi
> > ![2021-07-21_23-50](https://user-images.githubusercontent.com/79562439/126499822-922d7f31-94a7-49b5-94a0-1a7aa16c9582.png)
> > [mesh_CZM_test.zip](https://github.com/idaholab/moose/files/6856203/mesh_CZM_test.zip)
> 
> yes the mesh should not already have duplicated nodes, otherwise the algorithm will not find neighbors. That is exactly what is happening in your case.
> 
> For your second question is a bit trickier but the same principle apply. One way of doing it is to use the BreakMeshByBlockGenerator `block_pair` option see [here](https://mooseframework.inl.gov/source/meshgenerators/BreakMeshByBlockGenerator.html). That would require careful design of domains id.
> 
> I guess what we should really do in your case is to implement a join mesh algorithm such that you could start from a mesh from which you have removed cohesive elements... If you fill this might be an important tool could you open a ticket requesting for it?

If your mesh already has duplicated nodes along the interface, there is no reason to use `BreakMeshByBlockGenerator` again to create the interface. 

For your second question, if you still want to use CZM, maybe another option is to implement a new `MeshGenerator` that uses a lower dimension mesh (i.e. fracture surface) to create the CZM interface instead of using the blocks. 

For your application, are you trying to model crack propagation or just model the discrete fracture surfaces?

**Reply by Traiwit** [reply]
Hi @jiangwen84 @arovinelli thank you for your reply

I am working on mine modelling in order to produce stress field, displacements, and damage (plastic strain) v time.
The faults within the mesh are pre-defined (Abaqus with cohesive elements).

is there a way to manually generate the interface between the two blocks with mesh that has duplicated nodes? or the only way to work around this is to merge them back and apply `BreakMeshByBlockGenerator`

It seems like `BreakMeshByBlockGenerator` can be split into 2 sub-algorithms (from the attached figure).

![image](https://user-images.githubusercontent.com/79562439/126509215-e0503924-2af6-4953-a2e5-6dccc9b1a54f.png)

**Reply by jiangwen84** [reply]
Yes, we need to start with a fully connected mesh and then apply `BreakMeshByBlockGenerator` to create the interface and reconstruct the connectivity. The reason for that is that we need to use the `neighbor` information from the original mesh in our `InterfaceCZM` kernel.

**Reply by Traiwit** [reply]
@jiangwen84 @arovinelli sorry gents, I still can't get over this 

I'm checking `BreakMeshByBlockGenerator.C` I see it does 3 things, 1) making new duplicated nodes 2) attach the nodes back to the elements and then 3) create the cohesive interface. I only need to extract the last part (3) from the script.

ie I want to create a script that can generate cohesive interface from any of the available data listed below:

What I have available from Abaqus mesh are: 
1. duplicated nodes (list of node pairs)
2. list of surface pairs
3. list of element pairs

![image](https://user-images.githubusercontent.com/79562439/126781209-1a4cce43-e1bb-4e32-8ee0-4e181360640f.png)

is it possible to use any of these in order to generate the cohesive interface? I'm not very good with C++ it would be very helpful if you could point out which section of `BreakMeshByBlockGenerator.C` I should take out?

Thank you very much, much appreciated.

Kind regards,
Traiwit

**Reply by jiangwen84** [reply]
@Traiwit The MOOSE's CZM uses interface kernel (DG like) to enforce the CZM constitutive law. To make interface kernel to work, the element needs to know the information of its neighbor. In `BreakMeshByBlockGenerator.C`, we duplicate nodes and create new boundary for interface. One trick we did there is to keep the original `element-neighbor` information.

 For your case, if you mesh already has disconnected interface, MOOSE won't know the `element-neighbor` information by directly  reading the mesh. All you need to do is  to manually set the neighbor information for the elements along the interface by using list of element pairs. There might not much staff that you can reuse from `BreakMeshByBlockGenerator.C` because the majority of that code is the node duplication.

**Reply by arovinelli** [reply]
To give you some background, we split the mesh on the fly without breaking the connectivity because we don't have to worry with element is a side neighbor of another element. In practice in MOOSE each element has a list of side neighbors.
Unfortunately, when you start from a disconnected mesh, that list is incomplete. 
What you would need to do is to identify pairs of element neighbors, include which side is neighboring which side, i.e. elem 5, side 3 is a neighbor of elem 7 on side 1. once you have that list than it should be trivial to reconstruct the connectivity, but the code doing that, i.e. assigning a side neighbor to an element, is not in `BreakMeshByBlock`. 
If reminder correctly once you have the list of [elem_id, elem_side, neighbor_id, neighbor_side] the command you should use is
```
Elem * current_elem = mesh->elem_ptr(elem_id)
Elem * neighbor_elem = mesh->elem_ptr(neighbor_id)
elem-> set_neighbor(elem_side, neighbor_elem->id())
neighbor_elem-> set_neighbor(neighbor_side, elem->id())
```
that should do it
the only annoyance is how to get the list...
you can do it in moose directly (safest way as when you import a mesh there is always the possibility elemntids get changes):
this would imply something like

1-loop over all the nodes find all duplicated nodes
2-construct a node-to-element-map adding adding to it only duplicated nodes
3-using the node to element map and duplicated nodes find what element shoudl be neighbors
4-use the code above to stitch them.

 unfortunately is not going to be as easy as writing it.

**Reply by Traiwit** [reply]
Hi @arovinelli,

I believe we can extract the list of [elem_id, elem_side, neighbor_id, neighbor_side] from Abaqus mesh.

But I'm still unsure how to convert this information into cohesive interfaces.

Kind regards,
Traiwit

**Reply by jiangwen84** [reply]
@Traiwit I feel the effort to convert your Abaqus mesh would be more than to create a new mesh and assign blocks. You can use [gmsh](https://gmsh.info/)  or [cubit](https://coreform.com/products/coreform-cubit/) to build a mesh.

**Comment by arovinelli** [comment, marked as answer]
@jiangwen84 you need to reconstruct the connectivity somehow... as far as I know, there is not a generator available that start from a split mesh and reconstruct the connectivity or do we?

**Reply by jiangwen84** [reply]
> @jiangwen84 you need to reconstruct the connectivity somehow... as far as I know, there is not a generator available that start from a split mesh and reconstruct the connectivity or do we?

@arovinelli  Yes, we need to implement new generators. The current CZM interface generator uses blocks to define the interface location. A second option I suggest is to use lower dimension mesh to define the interface. But we need to make sure that lower dimension mesh sits on the element edges of the computation mesh. Either way we need reconstruct the connectivity.

**Reply by arovinelli** [reply]
> > @jiangwen84 you need to reconstruct the connectivity somehow... as far as I know, there is not a generator available that start from a split mesh and reconstruct the connectivity or do we?
> 
> @arovinelli Yes, we need to implement new generators. The current CZM interface generator uses blocks to define the interface location. A second option I suggest is to use lower dimension mesh to define the interface. But we need to make sure that lower dimension mesh sits on the element edges of the computation mesh. Either way we need reconstruct the connectivity.

If nodes are coincident wouldn't be easier simply to look for element with coincident faces but not sharing all the nodes? once we find the faces we simply add them to a boundary, still it would be difficult to discern between different surfaces...

**Comment by chunhuizhao478** [comment]
I think this question is relevant to our problem, so I would like to post my questions here: Is there a way for CZM to create/identify embedded cracks? I don't think the "BreakMeshByBlockGenerator" is able to handle this since there is no actual "block" between the crack interfaces. Or we need to get our hand dirty to pre-split, provide the list and write them in the code as suggested by arovinelli?

**Reply by hugary1995** [reply]
If you mean pre-existing embedded cracks, at the moment we only have two mesh generators for generating cohesive interfaces: `BreakMeshByBlockGenerator` and `ExplodeMeshGenerator`. You need to either predefine the subdomains so that you can use the existing mesh generators to generate the embedded cracks, or write your own CZM interface generator.

If you want to model actual crack nucleation and propagation, you'll want to use the `ExplodeMeshGenerator` and apply the CZM model everywhere, as every element-element pair is a potential nucleation site.

**Reply by chunhuizhao478** [reply]
Thanks for your reply. For the current stage we would like to introduce multiple pre-existing cracks. To use "BreakMeshByBlockGenerator", one needs to set different block-id to either side of the crack. For the case where it is not a through crack, what would be the best way to assign block-id? Could subdomains just surrounding the crack be easily identified?
![IMG_087E800A6807-1](https://user-images.githubusercontent.com/93891863/198153723-cd8693b0-e113-4212-81eb-7b4d52a457ab.jpeg)

**Reply by hugary1995** [reply]
Nice figure. How are you generating the mesh with the element edges aligned along the crack? If this mesh is generated using an external meshing tool, there should be an easy way to identify the blocks based on the starting and ending points of the pre-existing cracks.

**Reply by chunhuizhao478** [reply]
Thanks. If I use Gmsh, after creating the crack, I need to find a way to identify elements ID that surround the crack, then pass the information into MOOSE and assign different ID for elements on either side, then break the crack interface, is this correct?

**Reply by chunhuizhao478** [reply]
Btw, do you suggest using Gmsh or CUBIT as external mesh generator? Which one would be better?

**Reply by hugary1995** [reply]
> Thanks. If I use Gmsh, after creating the crack, I need to find a way to identify elements ID that surround the crack, then pass the information into MOOSE and assign different ID for elements on either side, then break the crack interface, is this correct?

That's correct, although I'm not sure how to conditionally assign a physical surface in gmsh, based on my limited experience with gmsh.

**Reply by hugary1995** [reply]
> Btw, do you suggest using Gmsh or CUBIT as external mesh generator? Which one would be better?

gmsh is free while cubit probably has more features. Your institution may be hosting a license of cubit.

**Reply by hugary1995** [reply]
BTW, generating a CZM interface based on starting and ending points of a line segment is also useful for us, so I may be able to write a mesh generator to do that later this year. You are welcome to create an issue to help me track that

**Reply by chunhuizhao478** [reply]
"there should be an easy way to identify the blocks based on the starting and ending points of the pre-existing cracks." Just want to make sure if there is another way, all I can think of right now is write scripts and play around in MSH file to get the information, but it is not easy to do.

**Reply by hugary1995** [reply]
It might be easier to use the gmsh python api.

**Reply by chunhuizhao478** [reply]
Oh yes, that's also something I'm looking into. Thanks!

**Reply by chunhuizhao478** [reply]
Just to update the question I asked, Gmsh + MESHIO can do the trick.

**Comment by chunhuizhao478** [comment]
I would like to ask for the reference article if possible for cohesive zone model implementation in moose (It is mentioned that the cohesive zone model is implemented using a discrete Galerkin approach).

**Reply by jiangwen84** [reply]
For small strain, it is very simple DG-like constraint. For large strain, this report https://www.osti.gov/biblio/1658575 is a good one. For mesh generator, see https://doi.org/10.1016/j.advengsoft.2014.04.002

**Reply by chunhuizhao478** [reply]
Thanks for your reply, for now we are dealing with small strain, but would like to gain more idea behind the model, could you direct me to similar DG-like constraint articles? Thanks

**Reply by jiangwen84** [reply]
Did you read the mesh generator paper? It seems that they have listed a few reference in their introduction section. If you cannot find what you need, let us know.

**Reply by chunhuizhao478** [reply]
Hi, I checked the mesh generator paper, but I'm not sure that is the same idea in the cohesive zone model implementation in MOOSE: The papers listed there are about hybrid DG/CZM methods, where "zero thickness interface element" is still introduced, but in the description of MOOSE, there is no cohesive element introduced in the formulation at all. So could you direct me to more specific articles where MOOSE adopts? Thanks!

**Reply by hugary1995** [reply]
I think if you introduce a discontinuity at an element edge (where this element is connected to its neighbor), that edge can be viewed as the so-called zero thickness element between this element and its neighbor.

**Reply by chunhuizhao478** [reply]
Thanks, these are very helpful. I'll look into the reference articles. 
To be more specific, the attempt to explode every element-element boundary as a CZM and study crack nucleation and propagation bring up some questions about its fundamentals: 
1. Whether the jump [[u]] is evaluated at quadrature point of interface element (plus/mins side) or nodal points?
2. How are the nodes sharing the same coordinate (after node duplication) interact with each other if one surface has experience tangential/normal jump and pass the motion forward to other surfaces? Take the three middle nodes in the figure for example.

<img width="326" alt="Screen Shot 2022-11-15 at 4 54 05 PM" src="https://user-images.githubusercontent.com/93891863/202042196-8c7d4643-45ce-4cf9-8451-a90870f95a29.png">

**Reply by hugary1995** [reply]
The jump is computed at qps on element edges. I don't quite get your 2nd question. Could you help me understand what you mean by "pass the motion forward to other surfaces"?

**Reply by chunhuizhao478** [reply]
Thanks, I’m just wondering if there is some continuity relations among the three nodes in the center so that if one surface has tangential jump, the other two surfaces can be affected (like a rupture propagates towards to other surfaces) or put it in the other way, how is cohesive zone model handle crack propagation in this case.

**Reply by chunhuizhao478** [reply]
Sorry if this may sound confusing, I try to understand how element cohesive zone surface interact with its neighbors

**Reply by Traiwit** [reply]
Hi guys, I still don't see the current MOOSE CZM and `BreakMeshByBlockGenerator` could handle more complex systems such as the one below:

![image](https://user-images.githubusercontent.com/79562439/202610184-cf2d056a-6283-4dbe-b855-206093207b81.png)

![image](https://user-images.githubusercontent.com/79562439/202610638-8e8451c4-4103-4622-8c0c-6c7b500f50f4.png)

These are C3D6 elements embedded within C3D4 mesh and connecting the node pair via NodalConstraint (works well with hydro/porousflow problems, but too expensive for mech problems)

I still believe introducing cohesive elements (zero thickness elements) still the way to go in the case that we already have pre-defined faults/cracks in the system.

Kind regards,
Traiwit

**Reply by jiangwen84** [reply]
IMO, the zero thickness and interface approaches are equivalent to handle cohesive conatraint. The issue for your problem is the preprocessing,  which is not trival for either approach.  BreakMeshByBlock is definitely not appropriate,  but some of its algorithm can be used to create interface boundaries,  e.g.  BreakMeshByLowerDMesh.
